File: examples/payment-service-go/payment.go
package payment

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

const (
	gatewayURL     = "https://gateway.example.com/charge"
	gatewayTimeout = 30 * time.Second
	maxRetries     = 3
	retryDelay     = 100 * time.Millisecond
)

var httpClient = &http.Client{Timeout: gatewayTimeout}

func callGateway(p *Payment) (string, error) {
	payload, err := json.Marshal(p)
	if err != nil {
		return "", fmt.Errorf("marshal payment: %w", err)
	}

	resp, err := httpClient.Post(gatewayURL, "application/json", bytes.NewReader(payload))
	if err != nil {
		return "", &PaymentError{Code: "network", Message: err.Error()}
	}
	defer resp.Body.Close()

	var data map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return "", fmt.Errorf("decode response: %w", err)
	}

	fmt.Println("DEBUG gateway response:", data)

	if resp.StatusCode != http.StatusOK {
		code := "card_declined"
		if resp.StatusCode == http.StatusPaymentRequired {
			code = "insufficient_funds"
		}
		return "", &PaymentError{Code: code, Message: fmt.Sprintf("gateway returned %d", resp.StatusCode)}
	}

	txID, ok := data["transactionId"].(string)
	if !ok {
		return "", fmt.Errorf("missing transactionId in response")
	}
	return txID, nil
}

func ProcessPayment(p *Payment) (string, error) {
	return WithRetry(func() (string, error) {
		return callGateway(p)
	}, RetryConfig{MaxAttempts: maxRetries, Delay: retryDelay})
}

func ValidatePayment(p *Payment) error {
	if p.Amount <= 0 {
		return fmt.Errorf("payment amount must be positive")
	}
	if p.CustomerID == "" {
		return fmt.Errorf("customer ID is required")
	}
	return nil
}

File: examples/payment-service-rust/src/payment.rs
use reqwest::Client;
use std::time::Duration;

use crate::retry::{with_retry, RetryConfig};
use crate::types::{Payment, PaymentError};

const GATEWAY_URL: &str = "https://gateway.example.com/charge";
const GATEWAY_TIMEOUT: Duration = Duration::from_secs(30);
const MAX_RETRIES: u32 = 3;
const RETRY_DELAY: Duration = Duration::from_millis(100);

async fn call_gateway(client: &Client, payment: &Payment) -> Result<String, PaymentError> {
    let response = client
        .post(GATEWAY_URL)
        .json(payment)
        .timeout(GATEWAY_TIMEOUT)
        .send()
        .await
        .map_err(|e| PaymentError::Network { message: e.to_string() })?;

    let status = response.status();

    let body: serde_json::Value = response
        .json()
        .await
        .map_err(|e| PaymentError::Network { message: e.to_string() })?;

    match status.as_u16() {
        200 => {
            let tx_id = body["transactionId"]
                .as_str()
                .ok_or_else(|| PaymentError::Network {
                    message: "missing transactionId in response".to_string(),
                })?;
            Ok(tx_id.to_string())
        }
        402 => Err(PaymentError::InsufficientFunds {
            message: format!("gateway returned {status}"),
        }),
        _ => Err(PaymentError::CardDeclined {
            message: format!("gateway returned {status}"),
        }),
    }
}

pub async fn process_payment(payment: &Payment) -> Result<String, PaymentError> {
    let client = Client::new();
    let config = RetryConfig {
        max_attempts: MAX_RETRIES,
        delay: RETRY_DELAY,
    };

    with_retry(&config, || call_gateway(&client, payment)).await
}

pub fn validate_payment(payment: &Payment) -> Result<(), PaymentError> {
    if payment.amount <= 0.0 {
        return Err(PaymentError::Network {
            message: "payment amount must be positive".to_string(),
        });
    }
    if payment.customer_id.is_empty() {
        return Err(PaymentError::Network {
            message: "customer ID is required".to_string(),
        });
    }
    Ok(())
}
